### PR TITLE
novnc backend

### PR DESCRIPTION
**Pull request by jcapona**
| Status  | Ticket/Issue |
| :---: | :--: |
| Ready/Hold | [Ticket](https://pi-top.atlassian.net/browse/FUR-2486) |

#### Main changes

Add backend to create and serve VNC over http(s) on further:
- Whenever a message is received via websockets with `'novnc': true`, a virtual display will be created (using `Xvfb`), a VNC server will publish it (using `x11vnc`) and will be accessible via web-browser (using `novnc`).
- Whenever activity is detected in a virtual display, a message will be sent to the frontend with connection details.
- When the received code stops, the virtual display is cleaned up.
- All of these tasks are performed by the `further-link-vnc` bash script.
- `novnc` serves the VNC server over https with a SSL certificate created on app start. 

#### Screenshots (feature, test output, profiling, dev tools etc)

[insert screenshots here]

#### Other notes (e.g. implementation quirks, edge cases, questions / issues)
-

#### Manual testing tips

- Install artifacts
- Send a command via websocket using the new `novnc` arg, such as:
```
{ "type": "start", "process": "1", "data": { "runner": "python3", "code": "print('hi'); from time import sleep; import turtle; turtle.Screen(); sleep(30); print('bye!')", "novnc": true } }
```

#### Tag anyone who definitely needs to review or help
-

### CODE REVIEW

File: further_link/endpoint/run.py [hunk 1]
```diff
@@ -116,11 +118,16 @@ async def on_output(channel, output):
             await self.send(channel, {"output": output}, process_id)
             logging.debug(f"{self.id} Sending Output {process_id} {channel} {output}")
 
+        async def on_display_activity(url):
+            logging.debug(f"{self.id} Sending display activity")
+            await self.send("novnc", {"url": url})
```
Comment (angusjfw) [inline further_link/endpoint/run.py:123]: `novnc` message will need the process_id

File: further_link/runner/process_handler.py [hunk 2]
```diff
@@ -194,6 +251,10 @@ async def _process_communicate(self):
         await asyncio.wait(self.ipc_tasks)
 
         await self._clean_up()
+        await self._clean_display()
+        if self.display_activity_monitor:
+            self.display_activity_monitor.stop()
```
Comment (angusjfw) [inline further_link/runner/process_handler.py:256]: i think this can all go in the main _clean_up method

File: further_link/runner/process_handler.py [hunk 3]
```diff
@@ -68,10 +76,47 @@ async def start(self, command, work_dir=None, env={}):
...
+
+            proc = await asyncio.create_subprocess_shell(
+                f"/usr/bin/further-link-vnc start {self.id}",
+                stdout=asyncio.subprocess.PIPE,
+                stderr=asyncio.subprocess.PIPE,
+            )
+
+            logging.info(f"Starting further-link-vnc on display '{self.display}'")
+            # Virtual display should be available before executing 'command'
+            await proc.wait()
```
Comment (angusjfw) [inline further_link/runner/process_handler.py:90]: we should probably log the command output here, at least as debug level

**Comment by angusjfw** [comment]
I know this isn't finished yet but I've been trying it out, it seems to work pretty great, and thought it might help to list the main things that seem to be missing:
- adding the dependencies xvfb, x11vnc, novnc
- installing the further-link-vnc script (also, i think maybe this should live in a scripts folder outside the python stuff like [this](https://github.com/pi-top/pi-top-4-Miniscreen/commit/1f8d4b7f8a7627f4a7f72dafaa88fb0374707e38)
- using the TLS certificates for the novnc server

File: further_link/endpoint/run.py [hunk 4]
```diff
@@ -116,11 +119,24 @@ async def on_output(channel, output):
...
             logging.debug(f"{self.id} Sending Output {process_id} {channel} {output}")
 
+        async def on_display_activity(url):
+            logging.debug(f"{self.id} Sending display activity")
+            parsed_url = urlparse(url)
+            await self.send(
+                "novnc",
+                {
+                    "port": parsed_url.port,
+                    "path": f"{parsed_url.path}/{parsed_url.query}",
```
Comment (angusjfw) [inline further_link/endpoint/run.py:129]:
```suggestion
                    "path": f"{parsed_url.path}?{parsed_url.query}",
```

File: further_link/runner/process_handler.py (hunk 3 above)
Comment (angusjfw) [reply further_link/runner/process_handler.py:90]: also, i think extract lines 82-90 into a method eg 'start_vnc' in utils/vnc.py

File: further_link/runner/process_handler.py [hunk 5]
```diff
@@ -68,10 +76,47 @@ async def start(self, command, work_dir=None, env={}):
...
+                proc = await asyncio.create_subprocess_shell(
+                    f"/usr/bin/further-link-vnc url {self.id}",
+                    stdout=asyncio.subprocess.PIPE,
+                    stderr=asyncio.subprocess.PIPE,
+                )
+                task = asyncio.create_task(proc.communicate())
+                await task
+                if task.done():
+                    novnc_url, _ = task.result()
+                    novnc_url = novnc_url.decode().strip()
```
Comment (angusjfw) [inline further_link/runner/process_handler.py:109]: another helper could go here, getting the url, and parsing it for the port and path
```
connection_details = await vnc_connection_details()
if connection_details:
    await self.on_display_activity(connection_details) # { port, path }
````

File: further_link/runner/process_handler.py [hunk 6]
```diff
@@ -126,6 +171,18 @@ async def stop(self):
...
+        if get_first_display() == self.display:
+            return
+
+        proc = await asyncio.create_subprocess_shell(
+            f"/usr/bin/further-link-vnc stop {self.display.replace(':', '')}",
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+        )
+        logging.info(f"Stopping further-link-vnc on display '{self.display}'")
+        asyncio.create_task(proc.communicate())
```
Comment (angusjfw) [inline further_link/runner/process_handler.py:184]: helper 'stop_vnc'

File: further_link/runner/process_handler.py [hunk 7]
```diff
@@ -216,6 +274,10 @@ async def _clean_up(self):
                 self.pty_slave.close()
                 os.remove(self.pty_master)
                 os.remove(self.pty_slave)
+            id_generator.free(self.id)
```
Comment (angusjfw) [inline further_link/runner/process_handler.py:277]: freeing the id should come after stopping the display which uses it

File: further_link/runner/process_handler.py (hunk 3 above)
Comment (angusjfw) [reply further_link/runner/process_handler.py:90]: maybe even keep starting activity monitor and providing connection details stuff inside the vnc helpers

`await start_vnc(self.id, self.on_display_activity)` - this could cover lines 82-113

**Review by angusjfw — COMMENTED** [review]
The test failure is legitimate, i'm seeing errors using the 'exec' runner:
```
Jun 08 16:39:18 pi-top further-link[12073]: INFO:aiohttp.access:192.168.178.30 [08/Jun/2022:15:39:18 +0000] "POST /upload HTTP/1.1" 200 303 "h>
Jun 08 16:39:18 pi-top further-link[12073]: INFO:root:Starting further-link-vnc on display '54'
Jun 08 16:39:19 pi-top further-link[12073]: ERROR:root:3021474208 Message Exception
Jun 08 16:39:19 pi-top further-link[12073]: Traceback (most recent call last):
Jun 08 16:39:19 pi-top further-link[12073]:   File "/usr/lib/python3/dist-packages/further_link/endpoint/run.py", line 64, in handle_message
Jun 08 16:39:19 pi-top further-link[12073]:     await self.add_handler(m_process, m_data["runner"], path, code, novnc)
Jun 08 16:39:19 pi-top further-link[12073]:   File "/usr/lib/python3/dist-packages/further_link/endpoint/run.py", line 139, in add_handler
Jun 08 16:39:19 pi-top further-link[12073]:     await handler.start(path, code, novnc=novnc)
Jun 08 16:39:19 pi-top further-link[12073]:   File "/usr/lib/python3/dist-packages/further_link/runner/exec_process_handler.py", line 34, in s>
Jun 08 16:39:19 pi-top further-link[12073]:     await super().start(command, work_dir, novnc=novnc)
Jun 08 16:39:19 pi-top further-link[12073]:   File "/usr/lib/python3/dist-packages/further_link/runner/process_handler.py", line 114, in start
Jun 08 16:39:19 pi-top further-link[12073]:     self.pgid = os.getpgid(self.process.pid)  # retain for cleanup
Jun 08 16:39:19 pi-top further-link[12073]: ProcessLookupError: [Errno 3] No such process
```

This doesn't seem to be directly related to your code changes, but I am finding the issue in this build only and not in the release.

The error occurs when you use the exec runner to run a script which ends very quickly (such as a bash `echo`, but not if you first do a short `sleep`). By the time we know the program has started and we go to call `os.getpgid` for it, it's already exited.

There is an existing, related issue that output is sometimes not captured when it comes very quickly after starting the program (eg much faster than python startup allows), which I made a [start](https://github.com/pi-top/Further-Link/tree/improve-output-handling) on improving a while ago.

Anyway, I have no idea why the behaviour has changed, it doesn't look related to code or even dependency changes, any ideas?

I'll add a commit to try-except this error from very quick exit and skip to the cleanup, it's a little awkward due to the way the pgid is used in the ipc system.

**Comment by angusjfw** [comment]
i think we can ignore the lgtm alert 'Clear-text storage of sensitive information'. i'm not sure if it's just a false positive or it's actually pretty smartly worked out that we've written sensitive data to a file

**Comment by angusjfw** [comment]
i'm pretty impressed lgtm bot has worked that one out 😁

**Review by angusjfw — APPROVED** [review]
